common/src/chainstate/utxo.rs

```diff
@@ -0,0 +1,462 @@
+            is_block_reward,
+            index: outpoint.get_output_index(),
+        }
+    }
+}
+
+impl From<&OutPointKey> for OutPoint {
+    fn from(key: &OutPointKey) -> Self {
+        let id = if key.is_block_reward {
+            let utxo_id: Id<Block> = Id::new(&key.outpoint_hash);
+            OutPointSourceId::BlockReward(utxo_id)
+        } else {
+            let utxo_id: Id<Transaction> = Id::new(&key.outpoint_hash);
+            OutPointSourceId::Transaction(utxo_id)
+        };
+
+        OutPoint::new(id, key.index)
+    }
+}
+
+/// The Unspent Transaction Output
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct Utxo {
+    output: TxOutput,
+    is_block_reward: bool,
+    /// At which height this containing tx was included in the active block chain
+    height: BlockHeight,
+}
+
+impl Utxo {
+    pub fn new(output: TxOutput, is_block_reward: bool, height: BlockHeight) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            height,
+        }
+    }
+
+    pub fn is_block_reward(&self) -> bool {
+        self.is_block_reward
+    }
+
+    pub fn height(&self) -> BlockHeight {
+        self.height
+    }
+
+    pub fn output(&self) -> &TxOutput {
+        &self.output
+    }
+
+    pub fn block_reward(&mut self, value: bool) {
+        self.is_block_reward = value;
+    }
+
+    pub fn set_height(&mut self, value: BlockHeight) {
+        self.height = value;
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct UtxoEntry {
+    /// None when utxo has been spent.
+    utxo: Option<Utxo>,
+    /// The utxo entry is dirty when this version is different from the parent.
+    is_dirty: bool,
+    /// The utxo entry is fresh when the parent does not have this utxo
+    is_fresh: bool,
+}
+
+impl UtxoEntry {
+    pub fn new(utxo: Utxo, is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: Some(utxo),
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    pub(in crate::chainstate) fn new_spent(is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: None,
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    fn utxo_mut(&mut self) -> Option<&mut Utxo> {
+        self.utxo.as_mut()
+    }
+
+    pub fn is_dirty(&self) -> bool {
+        self.is_dirty
+    }
+
+    pub fn is_fresh(&self) -> bool {
+        self.is_fresh
+    }
+
+    pub fn is_spent(&self) -> bool {
+        self.utxo.is_none()
+    }
+
+    pub fn utxo(&self) -> Option<Utxo> {
+        self.utxo.clone()
+    }
+}
+
+pub trait UtxosView {
+    /// Retrieves utxo.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo>;
+
+    /// Checks whether outpoint is unspent.
+    fn have_utxo(&self, outpoint: &OutPoint) -> bool;
+
+    /// Retrieves the block hash of the best block in this view
+    fn get_best_block_hash(&self) -> Option<H256>;
+
+    /// Estimated size of the whole view (0 if not implemented)
```

Comment:
Please use an `Option`. No special values like "0 if not implemented".

Reply:
option doesn't sit right with me... when it's asking for a size. That comment was copied from bitcoin, 
I updated the comment to "0 if empty".